swarm/network/stream/delivery.go

```diff
@@ -191,57 +172,171 @@ func (d *Delivery) Close() {
 	close(d.quit)
 }
 
-// RequestFromPeers sends a chunk retrieve request to a peer
-// The most eligible peer that hasn't already been sent to is chosen
-// TODO: define "eligible"
-func (d *Delivery) RequestFromPeers(ctx context.Context, req *network.Request) (*enode.ID, chan struct{}, error) {
-	requestFromPeersCount.Inc(1)
+// FindPeer is returning the closest peer from Kademlia that a chunk
+// request hasn't already been sent to
+func (d *Delivery) FindPeer(ctx context.Context, req *storage.Request) (*Peer, error) {
 	var sp *Peer
-	spID := req.Source
+	var err error
+
+	osp, _ := ctx.Value("remote.fetchh").(opentracing.Span)
+
+	depth := d.kad.NeighbourhoodDepth()
 
-	if spID != nil {
-		sp = d.getPeer(*spID)
-		if sp == nil {
-			return nil, nil, fmt.Errorf("source peer %v not found", spID.String())
+	originPo := -1
+	d.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
+		id := p.ID()
+
+		// get po between chunk and origin
+		if req.Origin.String() == id.String() {
+			originPo = po
+			return false
 		}
-	} else {
-		d.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
-			id := p.ID()
-			if p.LightNode {
-				// skip light nodes
-				return true
-			}
-			if req.SkipPeer(id.String()) {
-				log.Trace("Delivery.RequestFromPeers: skip peer", "peer id", id)
-				return true
+
+		return true
+	})
+
+	if osp != nil {
+		osp.LogFields(olog.Int("originPo", originPo))
+		osp.LogFields(olog.Int("depth", depth))
+	}
+
+	selectedPeerPo := -1
+
+	myPo := chunk.Proximity(req.Addr, d.kad.BaseAddr())
+
+	if osp != nil {
+		osp.LogFields(olog.Int("myPo", myPo))
+	}
+
+	skipped := 0
+
+	// do not forward requests if origin proximity is bigger than our node's proximity
+	// this means that origin is closer to the chunk
+	// 6 > 7
+	if originPo > myPo {
```

Comment:
```
The question here is: why would a node which is closer to the chunk have created a request to us (this node) in the first place? Can this happen at all?
```
NN

```
 If yes, is it correct in this (maybe edge) case to not forward the request?
```
Yes, because the `origin` will ask everyone in its NN. Your node doesn't have to do it for them.

Reply:
@holisticode this is not an edge case, it is a very valid base case.